Modules/_json.c

```diff
@@ -1503,6 +1503,13 @@ encoder_listencode_obj(PyEncoderObject *s, _PyAccu *acc,
             }
         }
         newobj = PyObject_CallFunctionObjArgs(s->defaultfn, obj, NULL);
+        if (obj != NULL && newobj != NULL && newobj->ob_type == obj->ob_type) {
```

Comment:
Thanks. I tried it but then when I ran the test suite, this test case failed with the error I have introduced.

    # Issue 16228: Crash on encoding resized list
    def test_encode_mutated(self):
        a = [object()] * 10
        def crasher(obj):
            del a[-1]
        self.assertEqual(self.dumps(a, default=crasher),
                 '[null, null, null, null, null]')